### PR TITLE
feat: QuarterlyNovember period type

### PR DESCRIPTION
**Pull request by abyot**

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/DateUnitPeriodTypeParser.java [hunk 1]
```diff
@@ -204,6 +204,37 @@ else if ( DateUnitType.QUARTERLY == dateUnitType )
 
             return new DateInterval( start, end );
         }
+        else if ( DateUnitType.QUARTERLY_NOV == dateUnitType )
+        {
+            int year = Integer.parseInt( matcher.group( 1 ) );
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/DateUnitPeriodTypeParser.java:209]: I believe that are too many "magic numbers" from lines 209 to 231.
Probably giving a name to them would make things easier to read/understand.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/DateUnitPeriodTypeParser.java [hunk 2]
```diff
@@ -204,6 +204,37 @@ else if ( DateUnitType.QUARTERLY == dateUnitType )
...
+            int quarter = Integer.parseInt( matcher.group( 2 ) );
+
+            // valid quarters are from 1 - 4
+            if ( quarter < 1 || quarter > 4 )
+            {
+                return null;
+            }
+
+            int monthOffset = -2;
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/DateUnitPeriodTypeParser.java:218]: Maybe we could have a method like "extractMonth()" where we can have a Javadoc explaining the calculation?
Also, giving a name for the numbers would make things easier to follow.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+        return periods;
+    }
+
+    /**
+     * Generates the last 4 quarters where the last one is the quarter which the
+     * given date is inside.
+     */
+    @Override
+    public List<Period> generateRollingPeriods( Date date )
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:114]: Would be nice to have, in the Javadoc, documentation for the params and the return object.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    // -------------------------------------------------------------------------
+    // CalendarPeriodType functionality
+    // -------------------------------------------------------------------------
+
+    @Override
+    public DateTimeUnit getDateWithOffset( DateTimeUnit dateTimeUnit, int offset, Calendar calendar )
+    {
+        return calendar.plusMonths( dateTimeUnit, offset * 3 );
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:82]: Could you add a short comment to tell why we multiply by 3 here? I know it's related to the quarter period, but it's just to make it more clear to anyone.

Also, would be nice to have, in the Javadoc, documentation for the params and the return object, and a short explanation for this method.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    /**
+     * Generates quarterly Periods for the whole year in which the given
+     * Period's startDate exists.
+     */
+    @Override
+    public List<Period> generatePeriods( DateTimeUnit dateTimeUnit )
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:90]: Would be nice to have, in the Javadoc, documentation for the params and the return object.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        dateTimeUnit.setMonth( 1 );
+        dateTimeUnit.setDay( 1 );
+
+        int year = dateTimeUnit.getYear();
+        List<Period> periods = Lists.newArrayList();
+
+        while ( year == dateTimeUnit.getYear() )
+        {
+            periods.add( createPeriod( dateTimeUnit, cal ) );
+            dateTimeUnit = cal.plusMonths( dateTimeUnit, 3 );
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:103]: Nice to have a short comment to tell why we add 3 months here.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    @Override
+    public List<Period> generateRollingPeriods( Date date )
+    {
+        date = createPeriod( date ).getStartDate();
+
+        return generateRollingPeriods( createLocalDateUnitInstance( date ), getCalendar() );
+    }
+
+    @Override
+    public List<Period> generateRollingPeriods( DateTimeUnit dateTimeUnit, Calendar calendar )
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:122]: Would be nice to have, in the Javadoc, documentation for the params and the return object.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    {
+        dateTimeUnit.setDay( 1 );
+
+        DateTimeUnit iterationDateTimeUnit = calendar.minusMonths( dateTimeUnit, 9 );
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:126]: Would be nice a short comment to tell why we subtract 9 months.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+        List<Period> periods = Lists.newArrayList();
+
+        for ( int i = 0; i < 4; i++ )
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:130]: Is related to quarterly? Could we add a comment or extract it into a meaningful variable?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            periods.add( createPeriod( iterationDateTimeUnit, calendar ) );
+            iterationDateTimeUnit = calendar.plusMonths( iterationDateTimeUnit, 3 );
+        }
+
+        return periods;
+    }
+
+    @Override
+    public Date getRewindedDate( Date date, Integer rewindedPeriods )
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java:140]: Would be nice to have, in the Javadoc, documentation for the params and the return object.
Explain what's happening and the months' calculations (including the multiplication).

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyNovemberPeriodType.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+     * n refers to the quarter, can be [1-4].
+     */
+    @Override
+    public String getIsoFormat()
+    {
+        return ISO_FORMAT;
+    }
+
+    @Override
+    public Period createPeriod( DateTimeUnit dateTimeUnit, Calendar calendar )
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyNovemberPeriodType.java:62]: Would be nice to explain the calculations through short comments or as a Javadoc.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyNovemberPeriodType.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    // PeriodType functionality
+    // -------------------------------------------------------------------------
+
+    @Override
+    public PeriodTypeEnum getPeriodTypeEnum()
+    {
+        return PeriodTypeEnum.QUARTERLY_NOV;
+    }
+
+    /**
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyNovemberPeriodType.java:52]: Probably, this comment should be on top of the constant `ISO_FORMAT`?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyPeriodType.java [hunk 13]
```diff
@@ -70,6 +57,15 @@ public PeriodTypeEnum getPeriodTypeEnum()
         return PeriodTypeEnum.QUARTERLY;
     }
 
+    /**
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyPeriodType.java:60]: Probably, this comment should be on top of the constant `ISO_FORMAT`?

File: dhis-2/dhis-api/src/test/java/org/hisp/dhis/period/QuarterlyNovemberPeriodTypeTest.java [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.List;
+
+import org.joda.time.DateTime;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * @author @author Abyot Asalefew Gizaw <abyota@gmail.com>
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/test/java/org/hisp/dhis/period/QuarterlyNovemberPeriodTypeTest.java:39]: Annotation `author` is duplicated.

File: dhis-2/dhis-web/dhis-web-commons-resources/src/main/webapp/dhis-web-commons/javascripts/dhis2/dhis2.period.js [hunk 15]
```diff
@@ -220,6 +220,7 @@ dhis2.period.PeriodGenerator = function(calendar, format) {
   this.registerGenerator( dhis2.period.MonthlyGenerator );
   this.registerGenerator( dhis2.period.BiMonthlyGenerator );
   this.registerGenerator( dhis2.period.QuarterlyGenerator );
```
Comment (maikelarabori) [inline dhis-2/dhis-web/dhis-web-commons-resources/src/main/webapp/dhis-web-commons/javascripts/dhis2/dhis2.period.js:222]: Should we have a frontend developer for the JS changes?

**Review by maikelarabori — COMMENTED** [review]
Hi @abyot, thanks for the changes. I left a few comments for review.
Also, don't we need a front-end developer for the Javascript changes...?

Cheers,
Maikel

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/DateUnitPeriodTypeParser.java (hunk 1 above)
Comment (abyot) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/DateUnitPeriodTypeParser.java:209]: have put a doc explaining the logic

**Pull request merged**